numpy/lib/twodim_base.py

```diff
@@ -565,7 +565,20 @@ def vander(x, N=None, increasing=False):
 
 def _histogram2d_dispatcher(x, y, bins=None, range=None, normed=None,
                             weights=None, density=None):
-    return (x, y, bins, weights)
+    yield x
+    yield y
+
+    # This terrible logic is adapted from the checks in histogram2d
+    try:
+        N = len(bins)
+    except TypeError:
+        N = 1
+    if N != 1 and N != 2:
```

Comment:
I think this should be `if N == 2:`, no? For `N>2`, presumably `bins` is already a single array_like, which will be used for both axes.

I wonder if it makes sense to port my test cases from astropy to numpy... see https://github.com/astropy/astropy/blob/11f12aae9b41395281dc4910dc6a7ec67c604dc8/astropy/units/tests/test_quantity_non_ufuncs.py from https://github.com/astropy/astropy/pull/8808

Reply:
(Though `histogram2d` has no coverage for `bins` yet - that was the reason for the issue...)

Reply:
The new code shows up green in the coverage check. Do you mean the [TestHistogramFunctions](https://github.com/astropy/astropy/blob/11f12aae9b41395281dc4910dc6a7ec67c604dc8/astropy/units/tests/test_quantity_non_ufuncs.py#L1209) ?

Reply:
I meant the whole lot...  Adapting it to a more minimal subclass (perhaps a fake `Quantity`). 

But it may well be much easier to rely on astropy and other projects that use `__array_function__` to do the testing.

Reply:
There are tests in astropy that failed before this PR and pass now?

Reply:
No, not yet, while writing tests I realized that there was a problem for `histogram*`, but I'm focussing on first getting the majority in. Adding further test cases is meant to be done as follow-up: https://github.com/astropy/astropy/issues/8825